### PR TITLE
Weight logic to Knockback

### PR DESCRIPTION
**Pull request by Vrixyz**
Related to https://github.com/ozkriff/zemeroth/issues/291

Impacts would only be "BombPush" ability used on ally Heavy agents.

Features:
- [x] Weight logic for BombPush
- [x] Same logic for FlyOff effect
- [x] Visualization for failed knockback
- [x] PR for assets/objects.ron definitions : https://github.com/ozkriff/zemeroth_assets/pull/6
- [x] Tests for Knockback
- [x] Tests for FlyOff

### CODE REVIEW

File: src/core/battle/effect.rs [hunk 1]
```diff
@@ -127,4 +127,5 @@ pub struct Dodge {
 pub struct Knockback {
     pub from: PosHex,
     pub to: PosHex,
+    // TODO: add strength there too ?
```
Comment (Vrixyz) [inline src/core/battle/effect.rs]: Adding strength here would allow to reduce distance based on weight ?

File: src/core/battle/execute.rs [hunk 2]
```diff
@@ -673,13 +673,15 @@ impl ExecuteContext {
...
     command: &command::UseAbility,
+    _: ability::Knockback,
 ) -> ExecuteContext {
     let mut context = ExecuteContext::default();
     let id = state::blocker_id_at(state, command.pos);
     let from = command.pos;
     let actor_pos = state.parts().pos.get(command.id).0;
     let dir = Dir::get_dir_from_to(actor_pos, command.pos);
     let to = Dir::get_neighbor_pos(command.pos, dir);
+    // Knockback strength might be used to push farther than 1 tile ?
```
Comment (Vrixyz) [inline src/core/battle/execute.rs]: I'm unsure if strength would mean :
- "distance to push" : strength of 1: pushes 1 tile ; strength of 3: pushes 3 tiles
- or "difficulty to resist": strength of 1 can be resisted by "Heavy" blockers ; strength of 3 cannot.

I assumed the first approach was preferred, without implementing different distances.

File: src/core/battle/tests.rs [hunk 3]
```diff
@@ -1256,7 +1257,12 @@ fn throw_bomb_push() {
         ),
         (
             "weak",
-            [component_agent_dull(), component_strength(1)].to_vec(),
+            [
+                component_agent_dull(),
+                component_strength(1),
+                component_blocker(Weight::Normal),
```
Comment (Vrixyz) [inline src/core/battle/tests.rs]: I should add a push test for heavy blockers :)

File: src/core/battle/execute.rs (hunk 2 above)
Comment (ozkriff) [reply src/core/battle/execute.rs]: I was thinking only about "difficulty to resist" for now.

File: src/core/battle.rs [hunk 4]
```diff
@@ -43,6 +46,36 @@ pub struct Id(i32);
...
+       }
+    }
+}
+
+#[derive(Serialize, Deserialize, Default, Clone, Copy, Debug, PartialEq, PartialOrd)]
+pub struct PushStrength(pub Weight);
+
+impl PushStrength {
+    pub fn can_push(self, weight: Weight) -> bool {
+        weight != Weight::Immovable && self.0 <= weight
```
Comment (Vrixyz) [inline src/core/battle.rs]: Adding logic in this file might not be what's best for the architecture, but I didn't know where to put it.

File: src/core/battle/state/apply.rs [hunk 5]
```diff
@@ -339,9 +339,13 @@ fn apply_effect_wound(state: &mut State, id: Id, effect: &effect::Wound) {
...
     assert!(state.map().is_inboard(effect.from));
     assert!(state.map().is_inboard(effect.to));
+
     assert!(!state::is_tile_blocked(state, effect.to));
     let parts = state.parts_mut();
-    parts.pos.get_mut(id).0 = effect.to;
+    if effect.strength.can_push(parts.blocker.get(id).weight) {
+        parts.pos.get_mut(id).0 = effect.to;
+    }
+    // else show a resistance ?
```
Comment (Vrixyz) [inline src/core/battle/state/apply.rs]: I havn't looked into graphics yet, I guess resistance feedback will be in another pull request

File: src/core/battle/tests.rs [hunk 6]
```diff
@@ -1256,13 +1257,28 @@ fn throw_bomb_push() {
...
+            ]
+            .to_vec(),
         ),
         ("bomb_push", Vec::new()),
     ]);
     let scenario = scenario::default()
         .object(P0, "thrower", PosHex { q: 0, r: 0 })
-        .object(P1, "weak", PosHex { q: 0, r: 3 });
+        .object(P1, "weak", PosHex { q: 0, r: 3 })
+        .object(P1, "heavy", PosHex { q: 1, r: 2 });
```
Comment (Vrixyz) [inline src/core/battle/tests.rs]: Sooo, I added a heavy object to the test, I'm not sure I've understood all the "Ids", but I guess it has to do with order of coordinates checking before applying events, It looks like it's deterministic so it's good for the tests :+1:

File: src/core/battle/tests.rs [hunk 7]
```diff
@@ -1314,22 +1330,32 @@ fn throw_bomb_push() {
...
                 .into(),
-                actor_ids: vec![Id(2)],
+                actor_ids: vec![Id(3)],
                 instant_effects: vec![
                     (
                         Id(1),
                         vec![effect::Knockback {
                             from: PosHex { q: 0, r: 3 },
                             to: PosHex { q: 0, r: 4 },
+                            strength: PushStrength { 0: Weight::Normal },
```
Comment (Vrixyz) [inline src/core/battle/tests.rs]: push strength is hardcoded in `push_bomb` ability.. ; I notice this test doesn't test the actual moving of the pushed entity, so it's a bit useless to test a "strong" force against normal/heavy/immovable

**Comment by Vrixyz** [comment]
Ok tests are not passing now, there must be something not deterministic in there, sometimes is succeeds and sometimes not, I guess the simplest fix is to split the test in 2 : "heavy" and "normal"

File: src/core/battle/tests.rs [hunk 8]
```diff
@@ -1326,6 +1332,106 @@ fn throw_bomb_push() {
...
+    let prototypes = prototypes(&[
+        (
+            "thrower",
+            vec![
+                component_agent_one_attack(),
+                component_abilities(&[ability_throw_bomb_push()]),
+            ],
+        ),
+        (
+            "heavy",
```
Comment (ozkriff) [inline src/core/battle/tests.rs]: hmm, it's confusing that this prototype is named "heavy" but its weight is "normal".

File: src/core/battle/tests.rs [hunk 9]
```diff
@@ -1326,6 +1332,106 @@ fn throw_bomb_push() {
...
+                    id: Id(2),
+                    pos: PosHex { q: 0, r: 2 },
+                    ability: Ability::ExplodePush,
+                }
+                .into(),
+                actor_ids: vec![Id(2)],
+                instant_effects: vec![
+                    (
+                        Id(1),
+                        vec![effect::Knockback {
```
Comment (ozkriff) [inline src/core/battle/tests.rs]: It might be a good idea to post-check the resulting position of the object. Something like

```rust
assert_eq!(state.parts().pos.get(Id(1)).0, PosHex { q: 2, r: 2 });
```

File: src/core/battle/state/apply.rs [hunk 10]
```diff
@@ -339,9 +339,13 @@ fn apply_effect_wound(state: &mut State, id: Id, effect: &effect::Wound) {
 fn apply_effect_knockback(state: &mut State, id: Id, effect: &effect::Knockback) {
     assert!(state.map().is_inboard(effect.from));
     assert!(state.map().is_inboard(effect.to));
+
```
Comment (ozkriff) [inline src/core/battle/state/apply.rs]: Could you please remove this empty line? I have a "soft code style" (it's not written anywhere yet) that empty lines inside functions should be avoided.

**Comment by ozkriff** [comment]
Could you rebase on top of the master branch (to fix Travis checks), please?

**Comment by ozkriff** [comment]
Also, this PR should be paired with a PR to https://github.com/ozkriff/zemeroth_assets/blob/master/objects.ron otherwise the actual game crashes during initialization with a 'Can't parse the prototypes: Message("missing field \`weight\`")' msg.

File: src/core/battle/tests.rs (hunk 8 above)
Comment (Vrixyz) [reply src/core/battle/tests.rs]: Indeed, the weight was not of a correct value, I'll update it to be "heavy"

File: src/core/battle/state/apply.rs (hunk 5 above)
Comment (Vrixyz) [reply src/core/battle/state/apply.rs]: Thanks, I implemented it

**Comment by ozkriff** [comment]
Added a few final touches.

Thanks for the PR again, @Vrixyz!!
